art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py

```diff
@@ -0,0 +1,450 @@
+            "batch_size": batch_size,
+            "random_eps": random_eps
+        }
+        ProjectedGradientDescentPytorch.set_params(self, **kwargs)
+
+        if self.random_eps:
+            lower, upper = 0, eps
+            mu, sigma = 0, (eps / 2)
+            self.norm_dist = truncnorm((lower - mu) / sigma, (upper - mu) / sigma, loc=mu, scale=sigma)
+
+    def generate(self, x, y=None, **kwargs):
+        """
+        Generate adversarial samples and return them in an array.
+
+        :param x: An array with the original inputs.
+        :type x: `np.ndarray`
+        :param y: Target values (class labels) one-hot-encoded of shape `(nb_samples, nb_classes)` or indices of shape
+                  (nb_samples,). Only provide this parameter if you'd like to use true labels when crafting adversarial
+                  samples. Otherwise, model predictions are used as labels to avoid the "label leaking" effect
+                  (explained in this paper: https://arxiv.org/abs/1611.01236). Default is `None`.
+        :type y: `np.ndarray`
+        :param mask: An array with a mask to be applied to the adversarial perturbations. Shape needs to be
+                     broadcastable to the shape of x. Any features for which the mask is zero will not be adversarially
+                     perturbed.
+        :type mask: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
+        if self.random_eps:
+            ratio = self.eps_step / self.eps
+            self.eps = np.round(self.norm_dist.rvs(1)[0], 10)
+            self.eps_step = ratio * self.eps
+
+        y = check_and_transform_label_format(y, self.estimator.nb_classes)
+
+        if y is None:
+            # Throw error if attack is targeted, but no targets are provided
+            if self.targeted:
+                raise ValueError("Target labels `y` need to be provided for a targeted attack.")
+
+            # Use model predictions as correct outputs
+            targets = get_labels_np_array(self.estimator.predict(x, batch_size=self.batch_size))
+        else:
+            targets = y
+
+        mask = kwargs.get("mask")
+        if mask is not None:
+            # Ensure the mask is broadcastable
+            if len(mask.shape) > len(x.shape) or mask.shape != x.shape[-len(mask.shape):]:
+                raise ValueError("Mask shape must be broadcastable to input shape.")
+
+        adv_x_best = None
+        rate_best = None
+
+        for _ in range(max(1, self.num_random_init)):
+            adv_x = x.astype(ART_NUMPY_DTYPE)
+
+            # Compute perturbation with implicit batching
+            for batch_id in range(int(np.ceil(x.shape[0] / float(self.batch_size)))):
+                batch_index_1, batch_index_2 = batch_id * self.batch_size, (batch_id + 1) * self.batch_size
+                batch = x[batch_index_1:batch_index_2]
+                batch_labels = targets[batch_index_1:batch_index_2]
```

Comment:
Would it be now possible to use `torch.utils.data` for the batching? It could improve performance.

Reply:
No, we need numpy as input for those framework-specific attack.

Reply:
Done.